### PR TITLE
feat: implement support for Stargate V2 (StargateFacetV2 v1.0.0, ReceiverStargateV2 v.1.0.0)

### PR DESCRIPTION
**Pull request by 0xDEnYO**

### CODE REVIEW

File: src/Facets/StargateFacetV2.sol [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+    /// @param _bridgeData Data used purely for tracking and analytics
+    /// @param _stargateData Data specific to Stargate Bridge
+    function startBridgeTokensViaStargate(
+        ILiFi.BridgeData calldata _bridgeData,
+        StargateData calldata _stargateData
+    )
+        external
+        payable
+        nonReentrant
+        refundExcessNative(payable(msg.sender))
```
Comment (janjakubnanista) [inline src/Facets/StargateFacetV2.sol:56]: Isn't in this case preferrable to use `tx.origin`? Depends on the desired outcome of course, if the `msg.sender` is a contract that's unable to `receive()`, the refund will result in a rejected transaction.

File: test/solidity/Facets/StargateFacetV2.t.sol [hunk 2]
```diff
@@ -0,0 +1,576 @@
...
+
+contract StargateFacetV2Test is TestBaseFacet {
+    // EVENTS
+    event WhitelistUpdated(address[] whitelistedRouters, uint256[] values);
+    event Approval(address owner, address spender, uint256 value);
+    event PartnerSwap(bytes2 partnerId);
+    error InvalidAssetId(uint16 invalidAssetId);
+
+    // These values are for Mainnet
+    address internal constant WETH_ADDRESS =
```
Comment (janjakubnanista) [inline test/solidity/Facets/StargateFacetV2.t.sol:35]: Nitpick: looks unused

File: test/solidity/Facets/StargateFacetV2.t.sol [hunk 3]
```diff
@@ -0,0 +1,576 @@
...
+        0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2;
+    address internal constant STARGATE_POOL_NATIVE =
+        0x77b2043768d28E9C9aB44E1aBfC95944bcE57931;
+    address internal constant STARGATE_POOL_USDC =
+        0xc026395860Db2d07ee33e05fE50ed7bD583189C7;
+    address internal constant TOKEN_MESSAGING =
+        0x6d6620eFa72948C5f68A3C8646d58C00d3f4A980;
+    uint256 internal constant DST_CHAIN_ID = 137;
```
Comment (janjakubnanista) [inline test/solidity/Facets/StargateFacetV2.t.sol:43]: Nitpick: Looks unused

File: test/solidity/Facets/StargateFacetV2.t.sol [hunk 4]
```diff
@@ -0,0 +1,576 @@
...
+    uint32 internal constant DST_E_ID_USDC = 30184; // BAS
+    uint32 internal constant DST_E_ID_NATIVE = 30111; // OPT
+    uint16 internal constant ASSET_ID_USDC = 1;
+    uint16 internal constant ASSET_ID_USDT = 2;
```
Comment (janjakubnanista) [inline test/solidity/Facets/StargateFacetV2.t.sol:47]: Nitpick: Looks unused

File: test/solidity/Periphery/ReceiverStargateV2.t.sol [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+        vm.startPrank(USER_SENDER);
+        dai.transfer(address(receiver), 1000);
+        vm.stopPrank();
+
+        // pull token
+        vm.startPrank(USER_DIAMOND_OWNER);
+
+        receiver.pullToken(ADDRESS_DAI, payable(USER_RECEIVER), 1000);
+
+        assertEq(1000, dai.balanceOf(USER_RECEIVER));
```
Comment (ryandgoulding) [inline test/solidity/Periphery/ReceiverStargateV2.t.sol:65]: I would add a test for native for sanity sake too:

```
vm.startPrank(USER_SENDER);
(bool success, ) = USER_SENDER.call{ value: 1000 }("");
vm.stopPrank();
assertTrue(success);
assertEq(0, address(USER_RECEIVER).balance);

vm.startPrank(USER_DIAMOND_OWNER);
receiver.pullToken(address(0), payable(USER_RECEIVER), 1000);
assertEq(1000, address(USER_RECEIVER).balance);
```

File: test/solidity/Periphery/ReceiverStargateV2.t.sol [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+        vm.startPrank(address(this));
+        vm.expectRevert(UnAuthorized.selector);
+
+        receiver.lzCompose{ gas: 400000 }(
+            address(0),
+            "",
+            composeMsg,
+            address(0),
+            ""
+        );
```
Comment (ryandgoulding) [inline test/solidity/Periphery/ReceiverStargateV2.t.sol:184]: Lines 175-184 are identical to Lines 163-173.  Consider just updating the comment in line 163 to say "deployer/owner"?

File: src/Periphery/ReceiverStargateV2.sol [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+    function lzCompose(
+        address _from,
+        bytes32, // _guid (not used)
+        bytes calldata _message,
+        address, // _executor (not used)
+        bytes calldata // _extraData (not used)
+    )
+        external
+        payable
+        //nonReentrant // @REVIEWER: pls comment/confirm that we do not need this here (due to onlyEndpointV2 access control modifier)
```
Comment (ryandgoulding) [inline src/Periphery/ReceiverStargateV2.sol:92]: Reentrancy is checked in EndpointV2 itself.  In fact, this is Testable, although it is a bit difficult to do so.  Here is a rough example of how this is tested:

https://github.com/lifinance/contracts/compare/lf-8443-starcraftV2...ryandgoulding:contracts:lf-8443-starcraftV2__feedback?expand=1

Essentially, the above test delivers native eth to a contract with a `receive()` that has been contrived to make the re-entrant `lzCompose()` call.  It is then checked that the call fails with the reason LZ_COMPOSE_NOT_FOUND_SELECTOR and that the hash has already been seen before.

File: test/solidity/Periphery/ReceiverStargateV2.t.sol [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+    }
+
+    function test_canDecodeStargatePayloadAndExecuteSwap() public {
```
Comment (ryandgoulding) [inline test/solidity/Periphery/ReceiverStargateV2.t.sol:198]: May be more interesting to call `EndpointV2.lzCompose(...)` with a non-permissioned user, as anyone can actually make the call.  An rough example can be found here:

https://github.com/lifinance/contracts/compare/lf-8443-starcraftV2...ryandgoulding:contracts:lf-8443-starcraftV2__feedback?expand=1

File: src/Facets/StargateFacetV2.sol (hunk 1 above)
Comment (0xDEnYO) [reply src/Facets/StargateFacetV2.sol:56]: Thanks for the comment. We never use tx.origin as this has other undesired implications. 
Most contracts that interact with our contract will have a receive function.
Closing this conversation as it's a known issue

File: test/solidity/Facets/StargateFacetV2.t.sol (hunk 2 above)
Comment (0xDEnYO) [reply test/solidity/Facets/StargateFacetV2.t.sol:35]: correct, removed

File: test/solidity/Facets/StargateFacetV2.t.sol (hunk 3 above)
Comment (0xDEnYO) [reply test/solidity/Facets/StargateFacetV2.t.sol:43]: correct, removed. Thanks

File: test/solidity/Facets/StargateFacetV2.t.sol (hunk 4 above)
Comment (0xDEnYO) [reply test/solidity/Facets/StargateFacetV2.t.sol:47]: correct, removed, thanks

File: script/demoScripts/demoStargateV2Tx.ts [hunk 9]
```diff
@@ -0,0 +1,428 @@
...
+  'bytes32', // Transaction Id
+  'tuple(address callTo, address approveTo, address sendingAssetId, address receivingAssetId, uint256 fromAmount, bytes callData, bool requiresDeposit)[]', // Swap Data
+  'address', // Receiver
+]
+
+const FEE_LIBRARY_ABI = [
+  'function applyFeeView((address,uint32,uint64,uint64,bool,bool)) view returns (uint64)',
+]
+
+const VALID_EXTRA_OPTIONS_VALUE =
```
[the remark this answers is not in the document]
Comment (0xDEnYO) [reply script/demoScripts/demoStargateV2Tx.ts:47]: I was not aware of this. Thank you so much. Added it to our demo script.

File: src/Facets/StargateFacetV2.sol [hunk 10]
```diff
@@ -0,0 +1,167 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.17;
+
+import { ILiFi } from "../Interfaces/ILiFi.sol";
+import { IStargate, ITokenMessaging } from "../Interfaces/IStargate.sol";
+import { LibAsset } from "../Libraries/LibAsset.sol";
+import { LibDiamond } from "../Libraries/LibDiamond.sol";
```
[the remark this answers is not in the document]
Comment (0xDEnYO) [reply src/Facets/StargateFacetV2.sol:7]: correct, removed (thanks)

(Do you use some sort of utility/extension that points out unused stuff (in Solidity files) for you and that you could recommend?

File: src/Facets/StargateFacetV2.sol [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+import { ReentrancyGuard } from "../Helpers/ReentrancyGuard.sol";
+import { InformationMismatch, AlreadyInitialized, NotInitialized, ContractCallNotAllowed, InvalidCallData } from "../Errors/GenericErrors.sol";
```
[the remark this answers is not in the document]
Comment (0xDEnYO) [reply src/Facets/StargateFacetV2.sol:9]: correct, removed, thanks

File: test/solidity/Facets/StargateFacetV2.t.sol [hunk 12]
```diff
@@ -0,0 +1,576 @@
...
+        // set facet address in TestBase
+        setFacetAddressInTestBase(address(stargateFacetV2), "StargateFacetV2");
+
+        // update default bridgeData
+        bridgeData.bridge = "stargate";
+        bridgeData.minAmount = defaultUSDCAmount;
+
+        // prepare default StargateData
+        stargateData = StargateFacetV2.StargateData({
+            assetId: 1,
```
[the remark this answers is not in the document]
Comment (0xDEnYO) [reply test/solidity/Facets/StargateFacetV2.t.sol:113]: updated

File: test/solidity/Facets/StargateFacetV2.t.sol [hunk 13]
```diff
@@ -0,0 +1,576 @@
...
+            sendParams: IStargate.SendParam({
+                dstEid: 30150,
+                to: addressToBytes32(USER_RECEIVER),
+                amountLD: defaultUSDCAmount,
+                minAmountLD: 0,
```
[the remark this answers is not in the document]
Comment (0xDEnYO) [reply test/solidity/Facets/StargateFacetV2.t.sol:118]: updated

File: test/solidity/Facets/StargateFacetV2.t.sol [hunk 14]
```diff
@@ -0,0 +1,576 @@
...
+        uint256 _amountLD
+    ) internal pure returns (IStargate.SendParam memory sendParams) {
+        sendParams = IStargate.SendParam(
+            lzChainId,
+            addressToBytes32(_sender),
+            _amountLD,
+            (900 * _amountLD) / 1000,
+            "",
+            "",
+            OftCmdHelper.bus()
```
[the remark this answers is not in the document]
Comment (0xDEnYO) [reply test/solidity/Facets/StargateFacetV2.t.sol:557]: Yes this was leftover from the development process. I've removed it. Thank you.

File: test/solidity/Facets/StargateFacetV2.t.sol [hunk 15]
```diff
@@ -0,0 +1,576 @@
...
+        returns (address[] memory addresses, uint256[] memory values)
+    {
+        addresses = new address[](length);
+        addresses[0] = address(uniswap);
+        values = new uint256[](length);
+        values[0] = 1;
+    }
+
+    /// @dev Helper function to create commonly used SendParam for riding the bus.
+    function _createRideBusSendParams(
```
[the remark this answers is not in the document]
Comment (0xDEnYO) [reply test/solidity/Facets/StargateFacetV2.t.sol:527]: removed

File: src/Periphery/ReceiverStargateV2.sol [hunk 16]
```diff
@@ -0,0 +1,239 @@
...
+    /// @param _transactionId the transaction id associated with the operation
+    /// @param _swapData array of data needed for swaps
+    /// @param assetId token received from the other chain
+    /// @param receiver address that will receive tokens in the end
+    /// @param amount amount of token
+    /// @param reserveRecoverGas whether we need a gas buffer to recover
+    function _swapAndCompleteBridgeTokens(
+        bytes32 _transactionId,
+        LibSwap.SwapData[] memory _swapData,
+        address assetId,
```
[the remark this answers is not in the document]
Comment (0xDEnYO) [reply src/Periphery/ReceiverStargateV2.sol:149]: Very valid feedback, thanks for pointing this out. We use the term "assetId" throughout all our contracts for a "contractAddress". 
I've added a comment to make it clear that those are not to be confused.

File: test/solidity/Periphery/ReceiverStargateV2.t.sol (hunk 5 above)
Comment (0xDEnYO) [reply test/solidity/Periphery/ReceiverStargateV2.t.sol:65]: added

File: src/Periphery/ReceiverStargateV2.sol (hunk 7 above)
Comment (0xDEnYO) [reply src/Periphery/ReceiverStargateV2.sol:92]: awesome, I've added this test to our repo for reference and removed the nonReentrant modifier from the Receiver contract. Thank you for the feedback and the test case

File: test/solidity/Periphery/ReceiverStargateV2.t.sol (hunk 8 above)
Comment (0xDEnYO) [reply test/solidity/Periphery/ReceiverStargateV2.t.sol:198]: Thanks, I added it.

File: src/Facets/StargateFacetV2.sol [hunk 17]
```diff
@@ -0,0 +1,166 @@
...
+        } else {
+            // ERC20
+            // check current allowance to router
+            ERC20 sendingAsset = ERC20(_bridgeData.sendingAssetId);
+            uint256 currentAllowance = sendingAsset.allowance(
+                address(this),
+                routerAddress
+            );
+            // check if allowance is sufficient
+            if (currentAllowance < _bridgeData.minAmount) {
```
Comment (ezynda3) [inline src/Facets/StargateFacetV2.sol:122]: Why not just use `LibAsset::maxApproveERC20()` here? It does all of these checks anyway...

File: src/Facets/StargateFacetV2.sol [hunk 18]
```diff
@@ -0,0 +1,166 @@
...
+    /// @dev returns the pool-/routerAddress for a given asset(Id), reverts if no address can be found
+    function _getRouterAddress(
+        uint16 assetId
+    ) private returns (address routerAddress) {
+        // get the router-/pool address through the TokenMessaging contract
+        routerAddress = tokenMessaging.stargateImpls(assetId);
+        if (routerAddress == address(0)) revert InvalidAssetId(assetId);
+    }
+
+    receive() external payable {}
```
Comment (ezynda3) [inline src/Facets/StargateFacetV2.sol:165]: I don't think we need this as the diamond already has it.

**Review by ezynda3 — DISMISSED** [review]

File: src/Facets/StargateFacetV2.sol (hunk 17 above)
Comment (0xDEnYO) [reply src/Facets/StargateFacetV2.sol:122]: Cause this is a gas-optimized version of achieving a max approval.
Main differences: 

- no check if sendingAssetId is != address(0) (as LibUtil does)
- no check if spender is != address(0) (as LibUtil does)
- middle step to approve to 0 only if current allowance is non-0 (LibUtil will do it in every case)

File: src/Facets/StargateFacetV2.sol (hunk 18 above)
Comment (0xDEnYO) [reply src/Facets/StargateFacetV2.sol:165]: I added it so that the facet is also usable as standalone contract.
It doesnt hurt, right?

Comment (0xDEnYO) [reply src/Facets/StargateFacetV2.sol]: I have removed it

**Review by maxklenk — APPROVED** [review]

**Pull request merged**